Support Git in WSL

gerhardol opened the pull request:
Fixes #8172

## Proposed changes

For Git repos stored in WSL2 directories, execute the Git executable in WSL
where possible to improve performance. WSL Git commands are several
times faster when executed by the WSL Git executable instead of the
Git Extensions native (Windows) executable.

The paths internal to Git Extensions are always native (Windows).
Using the WSL executable requires that paths provided to WSL Git
are adjusted from/to Git Extensions native/internal (Windows) format to WSL, like
\\wsl$\Ubunto\repo to /repo and c:\repo to /mnt/c/repo.

The Git Extensions native (Windows) Git executable is still used in some situations:
* All handling and settings related to Git in Settings. This includes
display of Git version as well.
However, if the Git version is too old to be supported, the GitVersion
check added in #9680 will report that the version is too old.
* Custom merge implementation in FormResolveConflicts
* Custom difftool/mergetool, the tools are only extracted in Windows.
Default git-difftool is using the Git executable, which could be separate
for WSL Git.
* Some built-in plugins etc, ScriptRunner and FindLargeFiles always use
 native Git

Some additional known limitations:
* The dialog that offers "--force-with-lease" for push do not appear as the output is not redirected. (So the Push dialog appears and --force-with-lease has to manually selected).
* GitHub plugin do not use the Windows token

Note: All commands are not tried in WSL, there may be a few more changes needed.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

See additions below
![image](https://user-images.githubusercontent.com/6248932/139748280-2b9563b3-40fb-4a79-a5ee-cbd029fde062.png)

### After

![image](https://user-images.githubusercontent.com/6248932/139748142-714b0cba-3f3d-48b7-b7e4-a5b27cb5e98a.png)

WSL is shown separately in Git command log
![image](https://user-images.githubusercontent.com/6248932/139602445-51df27c3-55ca-4834-83e1-a0ebc65981bb.png)

## Test methodology <!-- How did you ensure quality? -->

Tests added for PathUtil.
For the few changed commands, something could be added. There are no current tests for the commands.

## Merge strategy

Rebase merge (PR submitter must change the commit message for the last commit).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -12,6 +12,7 @@ public static class PathUtil
     {
         private static readonly IEnvironmentAbstraction EnvironmentAbstraction = new EnvironmentAbstraction();
         private static readonly IEnvironmentPathsProvider EnvironmentPathsProvider = new EnvironmentPathsProvider(EnvironmentAbstraction);
+        private const string _wslPrefix = @"\\wsl$\";
```
```suggestion
        private const string WslPrefix = @"\\wsl$\";
```

gerhardol replied:
done

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +70,14 @@ public interface IGitModule
+        /// Convert native (Windows) path to the path for the Git executable
+        /// (this could be for instance WSL)
+        /// </summary>
+        /// <param name="path">ge.exe path to repo.</param>
```
What does this mean?

gerhardol replied:
To set inherit in GitModule
```suggestion
        /// <param name="path">Git Extensions native (Windows) path.</param>
```

gerhardol replied:
Refreshed the descriptions in general

RussKie commented on UnitTests/GitCommands.Tests/Helpers/PathUtilTest.cs:
```diff
@@ -231,6 +231,52 @@ public void IsWslPath(string path, bool expected)
+            PathUtil.GetWslDistro(path).Should().Be(expected);
+        }
+
+        [TestCase(@"\\wsl$\Ubuntu\work\..\GitExtensions\", "", @"\\wsl$\Ubuntu\work\..\GitExtensions\")]
+        [TestCase(@"C:\work\..\GitExtensions\", "", @"C:\work\..\GitExtensions\")]
+        [TestCase(@"work\..\GitExtensions\", "", @"work\..\GitExtensions\")]
+        public void GetRepoPath_default(string path, string wslDistro, string expected)
+        {
+            PathUtil.GetRepoPath(path, wslDistro).Should().Be(expected);
+        }
+
+        [TestCase(@"\\Wsl$\Ubuntu\work\..\GitExtensions\", "Ubuntu", @"/work/../GitExtensions/")]
+        [TestCase(@"\\wsl$\Ubuntu\work/../GitExtensions", "Ubuntu", @"/work/../GitExtensions")]
+        [TestCase(@"\\wsl$\Ubuntu-20.04\work\..\GitExtensions\", "Ubuntu-20.04", @"/work/../GitExtensions/")]
+        [TestCase(@"C:\work\..\GitExtensions\", "Ubuntu", @"/mnt/c/work/../GitExtensions/")]
+        [TestCase(@"work\..\GitExtensions\", "Ubuntu", @"work/../GitExtensions/")]
+        public void GetRepoPath_wsl(string path, string wslDistro, string expected)
+        {
+            PathUtil.GetRepoPath(path, wslDistro).Should().Be(expected);
+        }
+
+        [TestCase(@"\\wsl$/Ubuntu/work/../GitExtensions", "Ubuntu", @"//wsl$/Ubuntu/work/../GitExtensions")]
+        [TestCase(@"\\wsl$\Ubuntu-20.04\work\..\GitExtensions\", "Ubuntu", @"\\wsl$\Ubuntu-20.04\work\..\GitExtensions\")]
+        public void GetRepoPath_unexpected_usage(string path, string wslDistro, string expected)
+        {
+            PathUtil.GetRepoPath(path, wslDistro).Should().Be(expected);
+        }
```
Looking at inputs and outputs for these tests I see that in some cases we're chaining inputs to POSIX and in other tests we don't. From a caller perspective this feels very arbitrary.
I think we ought to normalise the output to be consistent.

gerhardol replied:
OK
These paths are intended for the Git executable, so normalizing is required.

gerhardol replied:
Changed this in general and removed som ToPosixPath() that are no longer needed.

RussKie commented on GitUI/HelperDialogs/FormResetAnotherBranch.cs:
```diff
@@ -93,7 +94,7 @@ private void Ok_Click(object sender, EventArgs e)
             }
 
-            var command = GitCommandHelpers.PushLocalCmd(gitRefToReset.CompleteName, _revision.ObjectId, Module.WorkingDir, ForceReset.Checked);
+            var command = GitCommandHelpers.PushLocalCmd(gitRefToReset.CompleteName, _revision.ObjectId, Module.GetRepoPath(Module.WorkingDir), ForceReset.Checked);
```
We need to clearly define and explain the rules when `GetRepoPath` must be used.

Also any reason not to compute and cache this property when we initialise a `GitModule`?

gerhardol replied:
GetRepoPath is required for paths that the Git executable will use.
Maybe rename to GetPathForGitExe()?

RussKie replied:
> GetRepoPath is required for paths that the Git executable will use.

Should this be pushed from UI layer into the GitCommands layer then? That is, instead of forcing callers to remember/know that paths must be encoded in a specific way, should callee (i.e. git commands, git executable, etc.) perform encoding when appropriate?

RussKie replied:
What I'm getting at - is "use WSL" something a consumer should care, or is it an infra setting that GitModule and Co should take care of based on the characteristics of a repo?

gerhardol replied:
This handles specific arguments only. 
An alternative would be to guess what are file arguments to transform when running the Git command (ths is what the wslgit wrapper does). The problem with that is that it is hard to find what is arguments. You may translate for instance searches for "C:\repo" and it is hard to handle where paths end if the paths contain spaces.

For the global "use WSL" configuration, the user will need to know about this as some settings will have to be done separately in WSL. But there should be no need to know details.

gerhardol replied:
Renamed the method, hopefully clearer now

RussKie reviewed (commented):
When we open a repo we can definitively determine whether a repo is a Windows repo or a WSL repo, right? Why do we need the checkbox? Is there a use case where a user wouldn't want to run a native git for WSL repo?

gerhardol commented:
> When we open a repo we can definitively determine whether a repo is a Windows repo or a WSL repo, right? Why do we need the checkbox? Is there a use case where a user wouldn't want to run a native git for WSL repo?

Only if there are issues in the implementation or if you have a very old Git version in your repo.
Some setups still use Ubuntu-16.04 for instance, not supported directly.
I do not think it is a big issue to not support such distros though, this is new functionality.
This could be a settings file only setting too.

If you have opened the WSL repo and flips this switch you have to reopen the repo for the settings to have effect, so if this can be deprecated I would be happier to not have to document this behavior.
(Should have stated in the description that doc need to be updated.)

RussKie commented:
> Only if there are issues in the implementation or if you have a very old Git version in your repo.
> Some setups still use Ubuntu-16.04 for instance, not supported directly.
> I do not think it is a big issue to not support such distros though, this is new functionality.
> This could be a settings file only setting too.

Ubuntu 16.04 LTS reach EOL in Apr 2021 (though its EOS is until Apr 2026). If not _truly_ necessary, I'd rather not complicate our implementations, and offer a simple config-free support, if possible.

vbjay commented:
Small concern. I set up a windows git config. Then I use wsl. Does it use
its own config seperate from all the stuff I set up in windows version?  If
so, we need to show this in git config settings.

On Sun, Oct 31, 2021, 8:02 PM Igor Velikorossov ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitUI/HelperDialogs/FormResetAnotherBranch.cs
> <https://github.com/gitextensions/gitextensions/pull/9702#discussion_r739895985>
> :
>
> > @@ -93,7 +94,7 @@ private void Ok_Click(object sender, EventArgs e)
>                  return;
>              }
>
> -            var command = GitCommandHelpers.PushLocalCmd(gitRefToReset.CompleteName, _revision.ObjectId, Module.WorkingDir, ForceReset.Checked);
> +            var command = GitCommandHelpers.PushLocalCmd(gitRefToReset.CompleteName, _revision.ObjectId, Module.GetRepoPath(Module.WorkingDir), ForceReset.Checked);
>
> What I'm getting at - is "use WSL" something a consumer should care, or is
> it an infra setting that GitModule and Co should take care of based on the
> characteristics of a repo?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9702#discussion_r739895985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMKCXKLM4XRQUXI425DUJXKKXANCNFSM5HCYQXWA>
> .
>

gerhardol commented:
> Small concern. I set up a windows git config. Then I use wsl. Does it use its own config separate from all the stuff I set up in windows version? If so, we need to show this in git config settings.

Settings in GE is only used for native (Windows) Git. There are a number of hardcoded access to .gitconfig etc in addition to that Settings in started globally without knowledge of WSL repos. I have four different distros but do not use Git in more than two of them, all should not be listed.

So WSL Git is a "secondary citizen", must be documented. I do not believe this is a big limitation.
Personally, I could live with just being able to compare, stage etc in GE and other operations from WSL command line.

gerhardol commented:
> Ubuntu 16.04 LTS reach EOL in Apr 2021 (though its EOS is until Apr 2026). If not _truly_ necessary, I'd rather not complicate our implementations, and offer a simple config-free support, if possible.

Removed the option, changed the text to refer to WSL.

gerhardol commented:
Superseded by #9720 (as well as some parts in #9718).

gerhardol commented:
Squashed and rebased to simplify review
Old commits available in branch `tmp/i8172-wsl-git`

Added a few adjustments

RussKie commented on GitCommands/Git/Commands/GitCommandHelpers.cs:
```diff
@@ -194,14 +194,14 @@ public static ArgumentString ResetCmd(ResetMode mode, string? commit = null, str
             return new GitArgumentBuilder("push")
             {
-                $@"""file://{repoDir.ToPosixPath()}""",
+                $@"""file://{repoDir}""",
```
Is is possible to add an debug assertion that the path is in the posix format? It's good that we updated the comments but there're no guards to ensure the path is in the expected format.

gerhardol replied:
Adding Debug.Assert

RussKie commented on GitCommands/Git/Executable.cs:
```diff
@@ -107,12 +110,12 @@ private sealed class ProcessWrapper : IProcess
                         StandardErrorEncoding = errorEncoding,
                         FileName = fileName,
-                        Arguments = arguments,
+                        Arguments = prefixArguments + arguments,
```
Use string interpolation
```suggestion
                        Arguments = $"{prefixArguments}{arguments}",
```

gerhardol replied:
ok

RussKie commented on GitCommands/Git/GitModule.cs:
```diff
@@ -47,18 +47,40 @@ public sealed class GitModule : IGitModule
         private readonly IRevisionDiffProvider _revisionDiffProvider = new RevisionDiffProvider();
         private readonly GetAllChangedFilesOutputParser _getAllChangedFilesOutputParser;
+
+        // executable may use "native" Git (required for UI as well as FormProcess) or Git in WSL
```
```suggestion
        // The executable may use "native" Git (required for UI as well as FormProcess) or Git in WSL
```

gerhardol replied:
ok

RussKie commented on GitCommands/Git/GitModule.cs:
```diff
@@ -130,26 +152,35 @@ static bool HasGitModulesFile(string path)
+        /// </summary>
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
+        public string GetPathForNative(string path) => PathUtil.GetNativePath(path, _wslDistro);
```
This must be added to `IGitModule` as well.

gerhardol replied:
I had planned to not add this to the interface, changing.

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
+        /// <param name="wslDistro">Name of the distro or empty for non WSL paths</param>
+        /// <returns>Posix path if not a WSL distro (Windows Git), WSL path for WSL Git.</returns>
+        public static string GetRepoPath(string? path, string wslDistro)
```
❗  The annotation is incorrect: if `path` is nullable - this returns `null` on line:198. 
All callers of this method assume the return isn't null, and subsequently may result in NREs.

gerhardol replied:
Changing to not return null

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <param name="wslDistro">Name of the distro or empty for non WSL paths.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
+        public static string GetNativePath(string path, string wslDistro)
```
Either incorrect nullable annotations, or absent null checks and redundant checks below.

gerhardol replied:
Changing to not return null

RussKie replied:
We either need to annotate the args as `string?` or add `_ = path ?? throw new ANE(nameof(path))`

gerhardol replied:
The args are not null and should not be nullable
I will add a null check for path still

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+        /// </summary>
+        /// <param name="path">Directory path.</param>
+        /// <returns>Name of the distro or empty for non WSL2 paths</returns>
+        public static string GetWslDistro(string path)
```
Either incorrect nullable annotations, or absent null checks.

gerhardol replied:
I believe this is OK, not called with null and not returning null.
If path is null, an empty string is returned still, so `string? path` is possible

RussKie replied:
We either need to annotate the return as `string?` or add `_ = path ?? throw new ANE(nameof(path))`

gerhardol replied:
if path is null (despite annotation), GetWslDistroLength() returns -1 and the method returns ""

RussKie replied:
So if `path` _can_ be `null`, then it must be annotated as `string? path`.
If there's no code path to return a `null` then the return is `string`, else we must annotate it as `string?`.

gerhardol replied:
If path is null, it is handled, but not expected why annotation should be string.

RussKie commented on GitUI/CommandsDialogs/FormPush.cs:
```diff
@@ -533,7 +533,8 @@ private bool HandlePushOnExit(ref bool isError, FormProcess form)
                     if (!form.ProcessArguments.Contains(" -f ") && !form.ProcessArguments.Contains(" --force"))
                     {
-                        Trace.Assert(form.ProcessArguments.StartsWith("push "), "Arguments should start with 'push' command");
+                        // Note that e.g. WSL may add other arguments prior to the actual command so push may not be first.
```
```suggestion
                        // Note that WSL may add other arguments prior to the actual command so "push" may not be first.
```

gerhardol replied:
ok

RussKie commented on GitUI/CommandsDialogs/FormPush.cs:
```diff
@@ -533,7 +533,8 @@ private bool HandlePushOnExit(ref bool isError, FormProcess form)
                     {
-                        Trace.Assert(form.ProcessArguments.StartsWith("push "), "Arguments should start with 'push' command");
+                        // Note that e.g. WSL may add other arguments prior to the actual command so push may not be first.
+                        Debug.Assert(form.ProcessArguments.Contains("push "), "Arguments should start with 'push' command");
```
Is the comment still valid?

gerhardol replied:
Yes, the extra arguments (see first comment) is mostly hidden for the user

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
+        /// <param name="wslDistro">Name of the distro or empty for non WSL paths</param>
+        /// <returns>Posix path if not a WSL distro (Windows Git), WSL path for WSL Git.</returns>
+        public static string GetRepoPath(string? path, string wslDistro)
```
```suggestion
        public static string GetRepoPath(string? path, string? wslDistro)
```

gerhardol replied:
ok

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <param name="wslDistro">Name of the distro or empty for non WSL paths.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
+        public static string GetNativePath(string path, string wslDistro)
```
```suggestion
        public static string GetNativePath(string? path, string? wslDistro)
```

gerhardol replied:
ok

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+        /// </summary>
+        /// <param name="path">Directory path.</param>
+        /// <returns>Name of the distro or empty for non WSL2 paths</returns>
+        public static string GetWslDistro(string path)
```
```suggestion
        public static string GetWslDistro(string? path)
```

gerhardol replied:
ok

RussKie commented on GitCommands/PathUtil.cs:
```diff
@@ -159,7 +160,93 @@ private static string ResolveRelativePath(string path, string relativePath)
+            int distroLen = GetWslDistroLength(path);
+            return distroLen <= 0 ? "" : path.Substring(WslPrefix.Length, distroLen);
+
+            static int GetWslDistroLength(string path)
```
```suggestion
            static int GetWslDistroLength(string? path)
```

gerhardol replied:
ok

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
+        /// </summary>
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
+        public string GetPathForNative(string path);
```
```suggestion
        public string GetPathForNative(string? path);
```

gerhardol replied:
ok

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
+        /// For WSL Git, the path will be adjusted.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
+        /// <returns>Posix path if Windows Git, WSL path for WSL Git.</returns>
```
Does it return an empty string for an invalid path or throws?

gerhardol replied:
Posix unless identified as wsl

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
+        /// Convert a path to native (Windows) format.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
```
Does it return an empty string for an invalid path or throws?

gerhardol replied:
Empty if null path, set path separators to Windows if wslDistro is empty/null, otherwise try to convert the known path formats in WSL to Windows

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
         bool IsDetachedHead();
 
+        /// <summary>
+        /// Convert the path for the Git executable.
+        /// For WSL Git, the path will be adjusted.
```
```suggestion
        ///  Returns the path to the Git executable. For WSL Git the path is adjusted.
```

gerhardol replied:
Rewrote the documentation, slightly different

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
+        /// Convert the path for the Git executable.
+        /// For WSL Git, the path will be adjusted.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
```
```suggestion
        /// <param name="path">The path as seen by the Git Extensions native (Windows) executable.</param>
```

gerhardol replied:
Rewrote the documentation, slightly different

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
+        /// For WSL Git, the path will be adjusted.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
+        /// <returns>Posix path if Windows Git, WSL path for WSL Git.</returns>
```
```suggestion
        /// <returns>The POSIX path for Windows Git, WSL path for WSL Git.</returns>
```

gerhardol replied:
Rewrote the documentation, slightly different

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
+        /// <returns>Posix path if Windows Git, WSL path for WSL Git.</returns>
+        public string GetPathForGitExec(string? path);
+
+        /// <summary>
+        /// Convert a path to native (Windows) format.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
```
```suggestion
        /// <summary>
        ///  Returns the path in Windows native format.
        /// </summary>
        /// <param name="path">The path as seen by the Git executable, possibly WSL Git.</param>
        /// <returns>The path in the Windows format with native file separators.</returns>
```

gerhardol replied:
Rewrote the documentation, slightly different

RussKie commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -70,6 +74,21 @@ public interface IGitModule
         /// <summary>Indicates HEAD is not pointing to a branch (i.e. it is detached).</summary>
         bool IsDetachedHead();
 
+        /// <summary>
+        /// Convert the path for the Git executable.
+        /// For WSL Git, the path will be adjusted.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
+        /// <returns>Posix path if Windows Git, WSL path for WSL Git.</returns>
+        public string GetPathForGitExec(string? path);
+
+        /// <summary>
+        /// Convert a path to native (Windows) format.
+        /// </summary>
+        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
+        /// <returns>Path in in Windows format with native file separators.</returns>
+        public string GetPathForNative(string path);
```
Would it makes sense to name these as follows?
```cs
        public string GetDistroPath(string? path);
        public string GetWindowsPath(string? path);
```

gerhardol replied:
I still have a dream to support more than windows so I prefer the current name.

RussKie replied:
When we get to support some other OS we can rename it, I promise :)

gerhardol replied:
I can change. Will also change GitExec names, to be consistent. (Most Native names/comments).

Will write something in the comments like:
`With Windows path the GitExtensions.exe native path format is meant, this should be renamed if GE is running native on other OSes.`

gerhardol replied:
Updated the naming to `GetWindowsPath` and `GetGitExecPath` (the last is more descriptive to me, the distro may use several Git executables).

RussKie commented:
:+1:

On Thu, 18 Nov 2021 at 22:18, Gerhard Olsson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Plugins/GitUIPluginInterfaces/IGitModule.cs
> <https://github.com/gitextensions/gitextensions/pull/9702#discussion_r752144007>
> :
>
> > +        /// <summary>
> +        /// Convert the path for the Git executable.
> +        /// For WSL Git, the path will be adjusted.
> +        /// </summary>
> +        /// <param name="path">Path as seen by the Git Extensions native (Windows) executable.</param>
> +        /// <returns>Posix path if Windows Git, WSL path for WSL Git.</returns>
> +        public string GetPathForGitExec(string? path);
> +
> +        /// <summary>
> +        /// Convert a path to native (Windows) format.
> +        /// </summary>
> +        /// <param name="path">Path as seen by the Git executable, possibly WSL Git.</param>
> +        /// <returns>Path in in Windows format with native file separators.</returns>
> +        public string GetPathForNative(string path);
>
> I can change. Will also change GitExec names, to be consistent. (Most
> Native names/comments).
>
> Will write something in the comments like:
> With Windows path the GitExtensions.exe native path format is meant, this
> should be renamed if GE is running native on other OSes.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9702#discussion_r752144007>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQQLHY2KIJSDTPAZWDUMTOI7ANCNFSM5HCYQXWA>
> .
>

andrej-schaefer commented:
Great feature thanks,
have anyone an idea how to store the passphrase for the SSH permanently

Tried a lot using keychain, but it seems when using git command with wsl it dont use the keychain 
workaround was to go with putty again by using wsl.localhost instead of wsl$ for the path
someone at the top have written "should it be detected automatically or be configured", here is a reason why it should stay configurable

maybe i missing some config, but was not able to find any solution for now

gerhardol commented:
See discussion in https://github.com/gitextensions/gitextensions/issues/10274#issuecomment-1302644406
(avoid discussions in closed PRs)

The complexity for this feature had to be limited somehow, I am sure that other can improve on this solution like handle #10173 better.
